pikelet/src/surface/grammar.lalrpop

```diff
@@ -39,11 +42,29 @@ pub Term: Term<&'input str> = {
     <term: ExprTerm> ":" <ty: Term> => Term::Ann(Box::new(term), Box::new(ty)),
 };
 
+LetBindings: Vec<(Range<usize>, &'input str, Term<&'input str>)> = {
+    <bindings: LetBindings> <binding: TermEntry> ";" => {
+        let mut bindings = bindings;
+        bindings.push(binding);
+        bindings
+    },
+    <binding: TermEntry> ";" => vec![binding],
+    <bindings: LetBindings> <binding: TypeEntry> ";" => {
+        let mut bindings = bindings;
+        bindings.push(binding);
+        bindings
+    },
+    <binding: TypeEntry> ";" => vec![binding],
+}
+
 ExprTerm: Term<&'input str> = {
     ArrowTerm,
     <start: @L> "fun" <param_names: ParamName+> "=>" <body: ExprTerm> => {
         Term::FunctionTerm(start.., param_names, Box::new(body))
     },
+    <start: @L> "let" <bindings: LetBindings> "in" <expr: ExprTerm> => {
```

Comment:
Could we have something like:
```
<start: @L> "let" <items: (<Item> ";")+> <last: Item?> "in" <expr: ExprTerm>
```

Items can either be declarations or definitions. You can see an example of this kind of thing in the way that struct entries are defined.

Reply:
I could have sworn that (I had both tried) and that it wouldn't work for this, because of the necessity of the final semi-colon in making nested lets unambiguous.

But something similar worked, I had to move items/last into a separate macro,
because we have to first accumulate them into a vector, that is we can't call the extend, and return a Term then later extend the term I believe.

Anyhow, I'm not sure why I couldn't get that working earlier :confused:

Reply:
I should probably rename it `Items<Item>`, rather than `Sequence<Item>`?
Btw, that extend trick from `List<entry>` :+1:

Reply:
Oooh looks good.

Yeah `Items` sounds good. Tbh I'm wondering if we should just use semicolons in lists and records, then we could reuse the same syntax!

Reply:
Almost the same syntax, we still need all the `List<entry> macro, because lists can be empty, so we're not any complexity savings by changing here it I believe, beyond that it is worth considering.

Reply:
Makes sense!